src/Parser/GuardianDruid/Modules/Features/DynamicHaste.js

```diff
@@ -0,0 +1,85 @@
+import Module from 'Parser/Core/Module';
+
+import Combatants from 'Parser/Core/Modules/Combatants';
+import Haste from 'Parser/Core/Modules/Haste';
+
+class DynamicHaste extends Module {
+  static dependencies = {
+    combatants: Combatants,
+    haste: Haste,
+  };
+
+  _hasteLog = [];
+  on_initialized() {
+    const baseHaste = this.combatants.selected.hastePercentage;
+    this.recordHasteChange(baseHaste, this.owner.fight.start_time);
+  }
+
+  on_changehaste(event) {
+    this.recordHasteChange(event.newHaste, this.owner.currentTimestamp);
+  }
+
+  recordHasteChange(newHaste, timestamp) {
+    if (this._hasteLog.length === 0) {
+      this._hasteLog.push({ timestamp, haste: newHaste });
+    }
+
+    const lastHasteIndex = this._hasteLog.length - 1;
+    const lastHaste = this._hasteLog[lastHasteIndex];
+
+    if (lastHaste.timestamp === timestamp) {
+      // If the haste change occured simultaneously with another haste change,
+      // treat it as a single change
+      this._hasteLog[lastHasteIndex].haste = newHaste;
+    }
+
+    if (lastHaste.haste !== newHaste) {
+      // Conversely, only record a haste change if the values are actually different
+      this._hasteLog.push({ timestamp, haste: newHaste });
+    }
+  }
+
+  isHastedAbilityOffCooldown(lastCast, timestamp, baseCD) {
```

Comment:
I'll soon finish and release the new cooldown module, I reckon it will be a little bit simpler to use. I think the only main thing left of the module is writing tests (and refactoring and fixing issues I run into), so you can already see it in the branch I linked earlier. Making it work with "hasted cooldowns" should be as easy as adjusting the cooldown on every `changehaste` event (I might ask you about how this works by the time I get to that). I hope you won't mind later migrating this to avoid having two modules doing the same.

Reply:
Yeah that's not a problem, and I'd be happy to help.

Basically the gist of it is instead of reducing the cooldown directly, I accelerate the time passed based on the current haste value, and I do it iteratively until the the cooldown ends or we reach the current time.  

I found this was slightly easier to reason about since I didn't need to hold a list of all spells and whether or not they were on cooldown, just a timeline of the haste buffs gained and lost.

Reply:
Yeah it's definitely a much easier way to do what you had to do. My approach is a lot more complicated at the base, but I think once it's done it should be pretty easy to change its behavior via the methods provided.